pageserver/src/tenant/timeline/layer_manager.rs

```diff
@@ -208,6 +208,19 @@ impl LayerManager {
         }
     }
 
+    /// LayerManager shutdown. The in-memory layers do cleanup on drop, so we must drop them in
+    /// order to allow shutdown to complete.
+    ///
+    /// If there was a want to flush in-memory layers, it must have happened earlier.
+    pub(crate) fn drop_inmemory_layers(&mut self, writer_state: &mut Option<TimelineWriterState>) {
+        let open = self.layer_map.open_layer.take();
+        let frozen = self.layer_map.frozen_layers.len();
+        self.layer_map.frozen_layers.clear();
+        assert_eq!(open.is_some(), writer_state.is_some());
+        writer_state.take();
+        tracing::info!(open = open.is_some(), frozen, "dropped inmemory layers");
+    }
```

Comment:
By simply taking out the in-memory layers, Timeline::get will be giving out wrong answers because we're not resetting last_record_lsn.

The only thing that prevents us from actually giving out the wrong answer is that we check for the cancellation token.
Too brittle for my taste.

So, I would much rather introduce a well-defined "we're shut down" state for the `LayerManager`, lik `enum LayerManager { Open(...), ShutDown }`.

Reply:
Timeline cancellation has been fired already and we take the write lock..

It think that should be enough.

Reply:
If we held the write lock on layer map until after the gate is closed an argument could be made that it's ok.

But, the reality is that we can always have places where we forget to check cancellation / gate, and if we access the layer map in those places, we return incorrect data.

I would sleep much better knowing there is robustness in depth.

Your call, I made my point.

Reply:
Yeah I do get your point. I'll try to convince myself more strongly. My expectation was that we would always have:

```
loop {
  // check cancellation
  // read_lock(search for remaining keys)
  // get_value_reconstruct_data(remaining keys, found layers)
  // maybe_continue
}
```

> defence in depth

I'd also like to prevent something else from getting worse: how the cancellations are spread around so wide right now. I am convinced we should not be checking for cancellations except for a very few places.

But as everything is so spread around maybe defense in depth is the way to go.

Reply:
There is at least the timeline_detail doing layer map accesses across the test suite after/at shutdown.

Reply:
Right now the many functions are not yet returning the Shutdown error, which would for example take precedence over RemoteTimelineClient's stopped error during late stages of compaction. Instead, there's `warn!` logging to give me (a lot of) test reports to verify.

Returning the `Result<(), Shutdown>` is no problem, and it is already handled even better than the current handling of the anyhow from RemoteTimelineClient.

Reply:
And I forgot to check before pushing... Well... Let's see the version with `Result<(), Shutdown>` then.